lib/core/api/get-file.js

```diff
@@ -24,43 +27,98 @@ module.exports = function(self) {
+    }
+
+    file._token = body.token;
+    file._id = body.id;
+    file._mimetype = body.mimetype;
+
+    getFilePointers(file, cb);
+  }
+
+  function getFilePointers(file, cb) {
+    var opts = {
+      token: file._token,
+      bucket: file._bucketId,
+      file: file._fileId
+    };
+
+    file._client.getFilePointers(opts, function(e, pointers) {
+      if(e) {
+        return _error(e);
+      }
+
+      resolveFileFromPointers(file, pointers, cb);
+    });
+  }
+
+  function resolveFileFromPointers(file, pointers, cb) {
+    file._client.resolveFileFromPointers(pointers, function(e, muxer) {
+      if(e) {
+        return _error(e);
+      }
+
+      file._muxer = muxer;
+      file.emit('ready');
+      saveFile(file);
+      return cb(null, file);
+    });
+  }
+
+  function saveFile(file) {
+    var stream = file._store.createWriteStream(file._fileId);
+    var key = KeyIV.getDeterministicKey(file._encryptionKey, file._fileId);
+    var secret = new KeyIV(key, file._fileId);
+    var decrypter = new DecryptStream(secret);
+    file._muxer.pipe(decrypter).pipe(stream);
+    file._muxer.on('error', file.emit.bind(file, 'error'));
+    stream.on('error', file.emit.bind(file, 'error'));
+    stream.on('data', file.emit.bind(file, 'data'));
+    file._muxer.on('end', function () {
+      stream.end();
+      delete file._muxer;
+      file.emit('done');
+    });
+  };
+
   /*
    * Get a file stream by bucketId and fileId
    * @param bucketId
    * @param fileId
   */
   return function(bucketId, fileId, cb){
+    var file = new self.File(self);
+    file._bucketId = bucketId;
+    file._fileId = fileId;
+    file._key = self._key.getPrivateKey();
+
+    // Start waterfall of above functions
+    createFileToken(file, cb);
+
     // First grab the the token for downloading the file
-    self.createFileToken(bucketId, function(e, body) {
-      if(body.encryptionKey === '' && self._key === undefined) {
-        if(cb) {
-          return cb(new Error('You must supply a decryption key for private buckets.'), null);
-        };
-        throw new Error('You must supply a decryption key for private buckets.');
-      };
-
-      if(body.encryptionKey !== '') {
-        self._encryptionKey = body.encryptionKey;
-      };
-
-      self._file = fileId;
-      self._bucketId = bucketId;
-      if(e) { return _error(e); }
-      self._token = body.token;
-      self._id = body.id;
-      self._mimetype = body.mimetype;
-
-      // If successful, try to resolve the file pointers
-      self.getFilePointers(function(e, pointers) {
-        if(e) { return _error(e); }
-        // We now have a reference to every piece of our file, let's assemble it!
-        self.resolveFileFromPointers(pointers, {}, function (e, muxer) {
-          if(e) { return _error(e); }
-          self._muxer = muxer;
-          self.emit('ready');
-          self.initStore(cb);
-        });
-      });
-    });
+    return file;
```

Comment:
need a way of checking for if cb is supplied then callback with file else return it